compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,271 @@
+  // memory slice.
+
+  // Get the copy size:
+  auto copyMemRefType = cast<MemRefType>(copy.getOperand(1).getType());
+  if (!memref::isStaticShapeAndContiguousRowMajor(copyMemRefType)) {
+    return rewriter.notifyMatchFailure(
+        copy, "Cannot proceed: copy to non-static or non-contiguous, "
+              "non-row major memref.");
+  }
+  auto rank = copyMemRefType.getRank();
+  SmallVector<OpFoldResult> tileSize(rank - 1, rewriter.getIndexAttr(1));
+
+  int64_t elementBitWidth = copyMemRefType.getElementTypeBitWidth();
+  if (kNumBitsPerCopy % elementBitWidth != 0) {
+    return rewriter.notifyMatchFailure(
+        copy, "Cannot proceed: preferred copy size is not a multiple of "
+              "element bit width.");
+  }
+  int64_t elementsPerCopy = kNumBitsPerCopy / elementBitWidth;
+
+  // divide the copy by subgroup:
+  assert(subgroupSize.size() == 1); // only 1-D
+  assert(workgroupSize[0] % subgroupSize[0] == 0);
+
+  int64_t numSubgroups = workgroupSize[0] / subgroupSize[0];
+  int64_t totalCopySize = getTotalSize(copyMemRefType.getShape());
+  int64_t totalCopySizePerSubgroup = totalCopySize / numSubgroups;
+  int64_t numCopiesPerThread =
+      (totalCopySizePerSubgroup / elementsPerCopy) / subgroupSize[0];
+  int64_t residualElements =
+      totalCopySizePerSubgroup % (subgroupSize[0] * elementsPerCopy);
+
+  LDBG("-- elementsPerCopy: " << elementsPerCopy << "\n");
+  LDBG("-- workgroupSize: " << workgroupSize[0] << "\n");
+  LDBG("-- numSubgroups: " << numSubgroups << "\n");
+  LDBG("-- totalCopySize: " << totalCopySize << "\n");
+  LDBG("-- totalCopySizePerSubgroup: " << totalCopySizePerSubgroup << "\n");
+  LDBG("-- numCopiesPerThread: " << numCopiesPerThread << "\n");
+  LDBG("-- residualElements: " << residualElements << "\n");
+
+  if (residualElements != 0) {
+    return rewriter.notifyMatchFailure(
+        copy, "Cannot proceed: cannot handle copying residual elements.");
+  }
+
+  // build For loop
+  SmallVector<Attribute> mapping;
+  for (int64_t i = 0, e = rank; i < e; ++i) {
+    unsigned mappingId = static_cast<unsigned>(gpu::MappingId::LinearDim0) + i;
+    mapping.push_back(gpu::GPUThreadMappingAttr::get(
+        context, static_cast<gpu::MappingId>(mappingId)));
+  }
+  std::reverse(mapping.begin(), mapping.end());
+
+  Value subgroupId = rewriter.create<gpu::SubgroupIdOp>(loc, nullptr);
+  Value laneId = rewriter.create<gpu::LaneIdOp>(loc, nullptr);
+
+  // compute number of loads per thread
+  auto sourceType = cast<MemRefType>(copy.getOperand(0).getType());
+  MemRefType localType = cast<MemRefType>(copy.getOutputs().front().getType());
+
+  auto getGlobalGatherIndex = [&](Value sgIdVal, Value lIdVal,
+                                  Value indVar) -> Value {
+    return rewriter.create<affine::AffineLinearizeIndexOp>(
+        loc, ValueRange{sgIdVal, indVar, lIdVal},
+        ArrayRef<int64_t>{totalCopySizePerSubgroup,
```

Comment:
This array is the shape of the linearization, not the strides, so you want `{numSubgroups, subgroupSize[0], elementsPerCopy}`

... Also, I'm dubious on the fact that you're only using subgroup size[0] here?

Reply:
You are right, I have then changed the linearization approach to be easier to understand.